### PR TITLE
Challenge: Convert a String to Uppercase

### LINKED ISSUE #16: Convert a String to Uppercase
**Program Title:** 
Case Converter

**Description:** 
Write a program that accepts a string input and converts every character in that string to uppercase. 

**Expected Outcome:** 
A string of entirely uppercase characters.

**Helpful resources:**
https://www.programiz.com/python-programming/methods/string/upper

**Pull Request No**
#22

### PR DESCRIPTION
**Pull request by NoorelsalamAlmakki**
---
name: solution review 
about: A template PR for code review with a checklist
---

## **Description:**

This pull request introduces a function `convert_string_to_uppercase` for solving the challenge of "String to Uppercase" #16  its corresponding unit tests to ensure robust functionality and maintain high code quality. The updates include:

## Behavior

The `convert_string_to_uppercase` function processes an input string and converts it to its uppercase version. Below is an overview of its behavior:

### **Input:**

 - Accepts any input of string type.

### **Processing:**

 - Validates that the input is a string.
 - Converts all characters in the string to uppercase.

### **Output:**

 - Returns the uppercase version of the input string.

### **Edge Cases Handled:**

 - Empty string: Returns an empty string since there's nothing to convert.
 - Strings with non-alphabetical characters (e.g., numbers, symbols): Leaves non-alphabetical characters unchanged.
 - Mixed case strings: Converts all lowercase letters to uppercase while preserving non-alphabetical characters.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by Husayn01** [comment]
@NoorelsalamAlmakki kindly pull the latest changes from the main branch to yours (Do this for each of your branch), and update your pull request checklist so I can proceed with the review.

**Comment by NoorelsalamAlmakki** [comment]
@Husayn01 I have updated the pull request and now it's ready for review.

File: solutions/convert_string_to_uppercase.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    Parameters:
+        - text (str): any string text.
+    Returns:
+        - upper_string (str): the uppercase version of the input text.
+    Raises:
+        - AssertionError: if the input text is not a string.
+    Example:
+    >>> convert_string_to_uppercase("hello")
+    'HELLO'
+    >>> sum_of_digits("HELLo")
```
Comment (Husayn01) [inline solutions/convert_string_to_uppercase.py]: The doctests have errors. The reference to `sum_of_digits` is incorrect. Here is the correct function:
```suggestion
    >>> convert_string_to_uppercase("HELLo")
```

File: solutions/convert_string_to_uppercase.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    'HELLO'
+    >>> sum_of_digits("")
```
Comment (Husayn01) [inline solutions/convert_string_to_uppercase.py]: Same goes for this:
```suggestion
    >>> convert_string_to_uppercase("")
```

File: solutions/tests/test_convert_string_to_uppercase.py
Comment (Husayn01) [inline solutions/tests/test_convert_string_to_uppercase.py:1]: Excellent job here! All code pass checks.

File: solutions/convert_string_to_uppercase.py (hunk 1 above)
Comment (NoorelsalamAlmakki) [reply solutions/convert_string_to_uppercase.py]: There must be some kind of confusion. This pull request is for "convert a string to uppercase" not the "sum of digits" challenge.

Comment (Husayn01) [reply solutions/convert_string_to_uppercase.py]: Yes, i am aware. Hence, it's why  you should replace `sum_of_digits("HELLo")` function in your code with `convert_string_to_uppercase("HELLo")`

File: solutions/tests/test_convert_string_to_uppercase.py
Comment (NoorelsalamAlmakki) [reply solutions/tests/test_convert_string_to_uppercase.py:1]: Thank you for your feedback!

File: solutions/convert_string_to_uppercase.py (hunk 1 above)
Comment (NoorelsalamAlmakki) [reply solutions/convert_string_to_uppercase.py]: Ohhh, good catch! I will modify it. Thanks alot

**Review by Husayn01 — APPROVED** [review]
Having passed all requirements in the checklist, your branch is now ready to be merged with the `main` branch.
